File: swift-bridge/Sources/AudioUnitBridge/AVAudioUnitTimePitch.swift
import AudioToolbox
import AVFAudio
import Foundation

func encodeAvAudioUnitTimePitch(_ effect: AVAudioUnitTimePitch) -> [String: Any] {
    var dict = encodeAvAudioUnitTimeEffect(effect)
    dict["rate"] = effect.rate
    dict["pitch"] = effect.pitch
    dict["overlap"] = effect.overlap
    return dict
}

@_cdecl("au_av_time_pitch_create")
public func au_av_time_pitch_create(
    _ outUnit: UnsafeMutablePointer<UnsafeMutableRawPointer?>,
    _ outErrorMsg: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>
) -> Int32 {
    outUnit.pointee = nil
    outErrorMsg.pointee = nil
    outUnit.pointee = retainBox(AVAudioUnitTimePitch())
    return AU_OK
}

@_cdecl("au_av_time_pitch_release")
public func au_av_time_pitch_release(_ ptr: UnsafeMutableRawPointer) {
    releaseBox(ptr, as: AVAudioUnitTimePitch.self)
}

@_cdecl("au_av_time_pitch_snapshot_json")
public func au_av_time_pitch_snapshot_json(_ ptr: UnsafeMutableRawPointer) -> UnsafeMutablePointer<CChar>? {
    jsonCString(encodeAvAudioUnitTimePitch(borrowBox(ptr, as: AVAudioUnitTimePitch.self)))
}

@_cdecl("au_av_time_pitch_get_rate")
public func au_av_time_pitch_get_rate(_ ptr: UnsafeMutableRawPointer) -> Float {
    borrowBox(ptr, as: AVAudioUnitTimePitch.self).rate
}

@_cdecl("au_av_time_pitch_set_rate")
public func au_av_time_pitch_set_rate(_ ptr: UnsafeMutableRawPointer, _ rate: Float) {
    borrowBox(ptr, as: AVAudioUnitTimePitch.self).rate = rate
}

@_cdecl("au_av_time_pitch_get_pitch")
public func au_av_time_pitch_get_pitch(_ ptr: UnsafeMutableRawPointer) -> Float {
    borrowBox(ptr, as: AVAudioUnitTimePitch.self).pitch
}

@_cdecl("au_av_time_pitch_set_pitch")
public func au_av_time_pitch_set_pitch(_ ptr: UnsafeMutableRawPointer, _ pitch: Float) {
    borrowBox(ptr, as: AVAudioUnitTimePitch.self).pitch = pitch
}

@_cdecl("au_av_time_pitch_get_overlap")
public func au_av_time_pitch_get_overlap(_ ptr: UnsafeMutableRawPointer) -> Float {
    borrowBox(ptr, as: AVAudioUnitTimePitch.self).overlap
}

@_cdecl("au_av_time_pitch_set_overlap")
public func au_av_time_pitch_set_overlap(_ ptr: UnsafeMutableRawPointer, _ overlap: Float) {
    borrowBox(ptr, as: AVAudioUnitTimePitch.self).overlap = overlap
}

@_cdecl("au_av_time_pitch_as_time_effect")
public func au_av_time_pitch_as_time_effect(_ ptr: UnsafeMutableRawPointer) -> UnsafeMutableRawPointer {
    retainBox(borrowBox(ptr, as: AVAudioUnitTimePitch.self) as AVAudioUnitTimeEffect)
}

File: src/av_audio_unit_time_pitch.rs
//! `AVAudioUnitTimePitch` wrappers.

use core::ffi::c_void;

use serde::Deserialize;

use crate::av_audio_unit::AvAudioUnit;
use crate::av_audio_unit_time_effect::{AvAudioUnitTimeEffect, AvAudioUnitTimeEffectInfo};
use crate::error::AuError;
use crate::ffi;
use crate::util::{status_result, take_json};

/// Snapshot of an `AVAudioUnitTimePitch`.
#[derive(Debug, Clone, Deserialize)]
pub struct AvAudioUnitTimePitchInfo {
    #[serde(flatten)]
    pub time_effect: AvAudioUnitTimeEffectInfo,
    pub rate: f32,
    pub pitch: f32,
    pub overlap: f32,
}

/// Owned handle to an `AVAudioUnitTimePitch`.
pub struct AvAudioUnitTimePitch {
    ptr: *mut c_void,
}

unsafe impl Send for AvAudioUnitTimePitch {}

impl Drop for AvAudioUnitTimePitch {
    fn drop(&mut self) {
        unsafe { ffi::au_av_time_pitch_release(self.ptr) };
    }
}

impl AvAudioUnitTimePitch {
    /// Create a time-pitch unit.
    pub fn new() -> Result<Self, AuError> {
        let mut unit_ptr = core::ptr::null_mut();
        let mut error_ptr = core::ptr::null_mut();
        let status = unsafe { ffi::au_av_time_pitch_create(&mut unit_ptr, &mut error_ptr) };
        status_result(status, error_ptr)?;
        Ok(Self { ptr: unit_ptr })
    }

    /// Snapshot the time-pitch metadata.
    pub fn info(&self) -> Result<AvAudioUnitTimePitchInfo, AuError> {
        let ptr = unsafe { ffi::au_av_time_pitch_snapshot_json(self.ptr) };
        take_json(ptr)
    }

    /// Current playback rate.
    pub fn rate(&self) -> f32 {
        unsafe { ffi::au_av_time_pitch_get_rate(self.ptr) }
    }

    /// Set the playback rate.
    pub fn set_rate(&self, value: f32) {
        unsafe { ffi::au_av_time_pitch_set_rate(self.ptr, value) };
    }

    /// Current pitch shift in cents.
    pub fn pitch(&self) -> f32 {
        unsafe { ffi::au_av_time_pitch_get_pitch(self.ptr) }
    }

    /// Set the pitch shift in cents.
    pub fn set_pitch(&self, value: f32) {
        unsafe { ffi::au_av_time_pitch_set_pitch(self.ptr, value) };
    }

    /// Current overlap amount.
    pub fn overlap(&self) -> f32 {
        unsafe { ffi::au_av_time_pitch_get_overlap(self.ptr) }
    }

    /// Set the overlap amount.
    pub fn set_overlap(&self, value: f32) {
        unsafe { ffi::au_av_time_pitch_set_overlap(self.ptr, value) };
    }

    /// Clone the base `AVAudioUnitTimeEffect` handle.
    pub fn as_time_effect(&self) -> AvAudioUnitTimeEffect {
        let ptr = unsafe { ffi::au_av_time_pitch_as_time_effect(self.ptr) };
        AvAudioUnitTimeEffect::from_raw(ptr)
    }

    /// Current bypass state.
    pub fn bypass(&self) -> bool {
        self.as_time_effect().bypass()
    }

    /// Set the bypass state.
    pub fn set_bypass(&self, value: bool) {
        self.as_time_effect().set_bypass(value);
    }

    /// Clone the base `AVAudioUnit` handle.
    pub fn as_av_audio_unit(&self) -> AvAudioUnit {
        self.as_time_effect().as_av_audio_unit()
    }
}
